internal/runtime/local/run.go

```diff
@@ -0,0 +1,121 @@
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+func (r *Runtime) runGadget(runner runtime.Runner, gadget gadgets.GadgetInstantiate, operatorPerGadgetParamCollection params.ParamsCollection, params params.Params) error {
+	log := runner.Logger()
+
+	// Create gadget instance
+	gadgetInstance, err := gadget.NewInstance(runner)
+	if err != nil {
+		return fmt.Errorf("instantiate gadget: %w", err)
+	}
+
+	// Deferring getting results and closing to make sure operators got their chance to clean up properly beforehand
+	defer func() {
+		if closer, ok := gadgetInstance.(gadgets.CloseGadget); ok {
+			log.Debugf("calling gadget.Close()")
+			closer.Close()
+		}
+		if results, ok := gadgetInstance.(gadgets.GadgetResult); ok {
+			res, err := results.Result()
+			log.Debugf("setting result")
+			runner.SetResult(res, err)
+		}
+	}()
+
+	// Install operators
+	err = runner.Operators().PreGadgetRun(runner, gadgetInstance, operatorPerGadgetParamCollection)
+	if err != nil {
+		return fmt.Errorf("starting operators: %w", err)
+	}
+	defer runner.Operators().PostGadgetRun()
+	log.Debugf("found %d operators", len(runner.Operators()))
+
+	if gadget.Type() == gadgets.TypeTraceIntervals {
+		// Enable interval pushes
+		log.Debugf("enabling snapshots")
+		runner.Columns().EnableSnapshots(runner.Context(), time.Second, 2)
+	}
+
+	// Set event handler
+	if setter, ok := gadgetInstance.(gadgets.EventHandlerSetter); ok {
+		log.Debugf("set event handler")
+		switch gadget.Type() {
+		default:
+			setter.SetEventHandler(runner.Columns().EventHandlerFunc(runner.Operators().Enrich))
+		}
+	}
+
+	// Set event handler for array results
+	if setter, ok := gadgetInstance.(gadgets.EventHandlerArraySetter); ok {
+		log.Debugf("set event handler")
+		switch gadget.Type() {
+		default:
+			setter.SetEventHandlerArray(runner.Columns().EventHandlerFuncArray(runner.Operators().Enrich))
+		case gadgets.TypeTraceIntervals:
+			setter.SetEventHandlerArray(runner.Columns().EventHandlerFuncSnapshot("main", runner.Operators().Enrich)) // TODO: "main" is the node
+		}
+	}
+
+	// Set event handler
+	if setter, ok := gadgetInstance.(gadgets.EventEnricherSetter); ok {
+		log.Debugf("set event enricher")
+		setter.SetEventEnricher(runner.Operators().Enrich)
+	}
+
+	if startstop, ok := gadgetInstance.(gadgets.StartStopAltGadget); ok {
+		log.Debugf("calling gadget.StartAlt()")
+		err := startstop.StartAlt()
+		if err != nil {
+			startstop.StopAlt()
+			return fmt.Errorf("run gadget: %w", err)
+		}
+		defer func() {
+			log.Debugf("calling gadget.StopAlt()")
+			startstop.StopAlt()
+		}()
+	} else if startstop, ok := gadgetInstance.(gadgets.StartStopGadget); ok {
+		log.Debugf("calling gadget.Start()")
+		err := startstop.Start()
+		if err != nil {
+			startstop.Stop()
+			return fmt.Errorf("run gadget: %w", err)
+		}
+		defer func() {
+			log.Debugf("calling gadget.Stop()")
+			startstop.Stop()
+		}()
+	}
```

Comment:
> I was also expecting an `else if run,ok:= gadgetInstance.(gadgets.RunGadget); ok {` instead of making snapshot gadgets implement `Start` (wrapper for `Run`) and `Stop` (empty).

That's only the case for `snapshot/process`, and I think I did that because I had to make it to produce the same JSON results as before (to get to status quo with this PR in terms of integration tests). We can certainly change/add/optimize interfaces as needed, maybe split Start/Stop in two and call whatever is necessary. However, I wouldn't add Run as alternative so Start - or something that is too similar to what we already have, keeping the interfaces as few as possible.

Reply:
> However, I wouldn't add Run as an alternative to Start - or something that is too similar to what we already have, keeping the interfaces as few as possible.
>
I would instead do it because: 

1. We are mixing two different concepts in the IG context. The `Start` is used for a gadget that starts and keeps running until the user stops it. While `Run` is a gadget that runs for the time it needs for gathering certain information and automatically stops when it is done.
2. I don't like that `Stop` function in the snapshot tracer code because it is out of context for that kind of gadget.  

https://github.com/inspektor-gadget/inspektor-gadget/blob/97450ff6b9aa16c6b6d8cdd029d5107e4730cc38/pkg/gadgets/snapshot/process/tracer/tracer.go#L290-L300

Reply:
You're right - if `Run()` has the signature that `RunCollector()` in that function has (and not the same as Start()), it should be handled that way and not use Start/Stop. I still think I did it this way (and postpone refactoring) because of the expected integration test results, but I'll re-check on that.

To recap: We have/need to have

1. Gadgets that use Start/Stop and possibly (but not necessarily) stream events between those
2. Gadgets that return ([]byte, err) in their own format (without `Event`) (via Result() ([]byte, error))
3. Gadgets that return ([]*T, err) - this would apply to snapshot/process mentioned above and be Run() ([]*T, error)

And we also have a mix of 1&2 (profile/block-io, which is already handled). Profile gadgets could also be a mix of 1&3, I guess.